dcos_launch/terraform.py

```diff
@@ -0,0 +1,100 @@
+    def __init__(self, config: dict, env=None):
+        self.config = config
+        self.init_dir = config.get('init_dir')
+        if not self.init_dir:
+            self.init_dir = 'terraform-init-' + uuid.uuid4().__str__()
+            self.config['init_dir'] = self.init_dir
+        self.init_dir = os.path.abspath(self.init_dir)
+        self.cluster_profile_path = os.path.join(self.init_dir, 'desired_cluster_profile.tfvars')
+        if env is None:
+            env = os.environ.copy()
+
+        if config['platform'] == 'azure':
+            # TODO
+            pass
+        elif config['platform'] == 'aws':
+            # TODO
+            pass
+        elif config['platform'] == 'gcp':
+            if 'GCE_CREDENTIALS' in env:
+                with open('gcp_service_creds.json', 'w') as file:
+                    file.write(env['GCE_CREDENTIALS'])
+                config['terraform_config']['gcp_credentials_key_file'] = os.path.abspath('gcp_service_creds.json')
+            elif 'GOOGLE_APPLICATION_CREDENTIALS' in env:
+                config['terraform_config']['gcp_credentials_key_file'] = env['GOOGLE_APPLICATION_CREDENTIALS']
+            else:
+                raise util.LauncherError(
+                    'MissingParameter', 'Either GCE_CREDENTIALS or GOOGLE_APPLICATION_CREDENTIALS must be set in env')
+
+            if 'gcp_project' not in config['terraform_config']:
+                config['terraform_config']['gcp_project'] = json.load(
+                    config['terraform_config']['gcp_credentials_key_file'])['project_id']
+        else:
+            raise util.LauncherError('InvalidPlatform', "allowed values: 'gcp', 'aws' and 'azure'")
+
+        # create terraform directory
+        if not os.path.exists(self.init_dir):
+            os.makedirs(self.init_dir)
+            module = 'github.com/dcos/{}?ref={}/{}'.format(
+                'terraform-dcos-enterprise' if config['dcos-enterprise'] else 'terraform-dcos',
+                TERRAFORM_DCOS_EE_VERSION if config['dcos-enterprise'] else TERRAFORM_DCOS_VERSION,
+                config['platform'])
+            util.run_subprocess('terraform init -from-module ' + module, self.init_dir)
+
+            with open(self.cluster_profile_path, 'w') as file:
+                for k, v in config['terraform_config'].items():
+                    file.write(k + ' = ')
+                    if type(k) is dict:
+                        file.write('<<EOF\n{}\nEOF\n'.format(yaml.dump(v)))
+                    else:
+                        file.write('"{}"\n'.format(v))
+
+    def create(self):
+        # TODO ssh key helpers
+        util.run_subprocess('terraform apply -auto-approve -var-file ' + self.cluster_profile_path, self.init_dir)
+        return self.config
+
+    def wait(self):
+        """ Nothing to do here because unlike the other launchers, create() runs and also waits for a subprocess to
+        finish instead of just sending an http request to a provider.
+        """
```

Comment:
Another way to do it would be not waiting for the subprocess in util.run_subprocess and call wait() on that subprocess here instead, but:
1. it would complexify things
2. the user would miss some of the output and details as long as he doesn't call the wait command
3. I don't see the benefit in separating wait and create here

Reply:
I agree with all those points and this no-op is a good call for this MVP. May want to make that change for the regular onprem launcher later